pypesto/ensemble/covariance_analysis.py

```diff
@@ -0,0 +1,318 @@
+import numpy as np
+from typing import Union
+
+from ..ensemble import Ensemble, EnsemblePrediction
+from .utils import get_prediction_dataset
+
+
+def get_covariance_matrix_parameters(ens: Ensemble):
+    """
+    Compute the covariance of ensemble parameters.
+
+    Parameters
+    ==========
+    ens:
+        Ensemble object containing a set of parameter vectors
+
+    Returns
+    =======
+    covariance_matrix:
+        covariance matrix of ensemble parameters
+    """
+
+    # call lowlevel routine using the parameter ensemble
+    return get_covariance_matrix_lowlevel(dataset=ens.x_vectors.transpose())
+
+
+def get_covariance_matrix_predictions(
+        ens: Union[Ensemble, EnsemblePrediction],
+        prediction_index: int = 0) -> np.ndarray:
+    """
+    Compute the covariance of ensemble predictions.
+
+    Parameters
+    ==========
+    ens:
+        Ensemble object containing a set of parameter vectors and a set of
+        predictions or EnsemblePrediction object containing only predictions
+
+    prediction_index:
+        index telling which prediction from the list should be analyzed
+
+    Returns
+    =======
+    covariance_matrix:
+        covariance matrix of ensemble predictions
+    """
+
+    # extract the an array of predictions from either an Ensemble object or an
+    # EnsemblePrediction object
+    dataset = get_prediction_dataset(ens, prediction_index)
+
+    # call lowlevel routine using the prediction ensemble
+    return get_covariance_matrix_lowlevel(dataset=dataset)
+
+
+def get_spectral_decomposition_parameters(
+        ens: Ensemble,
+        normalize: bool = False,
+        lower_absolute_cutoff: float = None,
+        lower_relative_cutoff: float = None,
+        inv_upper_absolute_cutoff: float = None,
+        inv_upper_relative_cutoff: float = None,
+        only_identifiable_directions: bool = False,
+        only_separable_directions: bool = False):
```

Comment:
Should this be considered a private method? If not, maybe the arguments could be simplified to e.g.
```python
cutoffs_separable = (0.1, None)
entry_point(cutoffs_separable=cutoffs_separable)
```
Then parse e.g.
```python
def entry_point(cutoffs_identifiable = (None, None), cutoffs_separable = (None, None):
    only_identifiable_directions = np.array(cutoffs_identifiable).any()
    only_separable_directions = np.array(cutoffs_separable).any()
    assert sum([only_identifiable_directions, only_separable_directions]) < 2  # or == 1?
    lower_absolute_cutoff, upper_absolute_cutoff = cutoffs_separable
    inv_lower_absolute_cutoff, inv_upper_absolute_cutoff = cutoffs_identifiable
```

Might make it easier for users then.

Reply:
And set default cutoffs to `-np.inf` and `np.inf` I guess (not sure, maybe this is already done?)

Reply:
I like your idea of parsing the inputs in the beginning. The default cutoffs are implicitly implemented already. Default values at the level of 1e-16 would make sense, as this is machine precision on most computers.

Reply:
Okay, added a check and most hopefully meaningful default parameters. One could collect thresholds in a tuple, but I'm against that, as I prefer more function parameters with clear names over slimming down function api... Although that's a style-decision, I have to admit